numpy/core/src/umath/ufunc_object.c

```diff
@@ -924,30 +935,67 @@ convert_ufunc_arguments(PyUFuncObject *ufunc,
     PyObject *obj;
 
     /* Convert and fill in input arguments */
+    int all_scalar = 1;
+    int any_scalar = 0;
+    int all_legacy = 1;
     for (int i = 0; i < nin; i++) {
         obj = PyTuple_GET_ITEM(full_args.in, i);
 
         if (PyArray_Check(obj)) {
-            PyArrayObject *obj_a = (PyArrayObject *)obj;
-            out_op[i] = (PyArrayObject *)PyArray_FromArray(obj_a, NULL, 0);
+            out_op[i] = (PyArrayObject *)obj;
+            Py_INCREF(out_op[i]);
+        }
```

Comment:
I find the twisting in and out a bit confusing; it might be clearer to just do the scalar check here and then slightly duplicate it below for the non-array case. But not a big deal.